pkg/api/controller.go

```diff
@@ -5234,3 +5234,126 @@ func (c *Controller) GetUsageReportSummary(w http.ResponseWriter, r *http.Reques
 	}
 	writeResponse(w, r, http.StatusOK, response)
 }
+
+func (c *Controller) CreateUserExternalPrincipal(w http.ResponseWriter, r *http.Request, body apigen.CreateUserExternalPrincipalJSONRequestBody, userID, principalID string) {
+	ctx := r.Context()
+	if c.isExternalPrincipalNotSupported(ctx) {
+		writeError(w, r, http.StatusNotImplemented, "Not implemented")
+		return
+	}
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.CreateUserExternalPrincipalAction,
+			Resource: permissions.UserArn(userID),
+		},
+	}) {
+		return
+	}
+
+	c.LogAction(ctx, "create_user_external_principal", r, "", "", "")
+
+	err := c.Auth.CreateUserExternalPrincipal(ctx, userID, principalID)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+	writeResponse(w, r, http.StatusCreated, nil)
+}
+
+func (c *Controller) DeleteUserExternalPrincipal(w http.ResponseWriter, r *http.Request, userID, principalID string) {
+	ctx := r.Context()
+	if c.isExternalPrincipalNotSupported(ctx) {
+		writeError(w, r, http.StatusNotImplemented, "Not implemented")
+		return
+	}
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.DeleteUserExternalPrincipalAction,
+			Resource: permissions.UserArn(userID),
+		},
+	}) {
+		return
+	}
+	c.LogAction(ctx, "delete_user_external_principal", r, "", "", "")
+	err := c.Auth.DeleteUserExternalPrincipal(ctx, userID, principalID)
+	if c.handleAPIError(ctx, w, r, err) {
+		return
+	}
+	writeResponse(w, r, http.StatusNoContent, nil)
+}
+
+func (c *Controller) GetUserExternalPrincipal(w http.ResponseWriter, r *http.Request, userID, principalID string) {
+	ctx := r.Context()
+	if c.isExternalPrincipalNotSupported(ctx) {
+		writeError(w, r, http.StatusNotImplemented, "Not implemented")
+		return
+	}
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.ReadUserAction,
```

Comment:
I don't remember our discussion but making sure that `ReadUserAction` is what we want here